src/api.ts

```diff
@@ -0,0 +1,637 @@
+${varName}
+#elseif(${varName}.class.name === 'java.lang.Double')
+${varName}
+#elseif(${varName}.class.name === 'java.lang.Boolean')
+${varName}
+#elseif(${varName}.class.name === 'java.lang.LinkedHashMap')
+{
+#foreach(${itemVarName} in ${varName}.keySet())
+"${itemVarName}": ${this.jsonStage(
+      itemVarName,
+      level + 1
+    )}#if($foreach.hasNext),#end
+#end
+}
+#elseif(${varName}.class.name === 'java.util.ArrayList')
+[
+#foreach(${itemVarName} in ${varName})
+${this.jsonStage(itemVarName, level + 1)}#if($foreach.hasNext),#end
+#end
+]`.replace(/\n/g, `${new Array(level * 2).join(" ")}\n`);
+  }
+}
+
+function validateFunctionDecl(a: any): FunctionDecl {
+  if (isFunctionDecl(a)) {
+    return a;
+  } else if (isErr(a)) {
+    throw a.error;
+  } else {
+    throw Error("Unknown compiler error.");
+  }
+}
+
+/**
+ * Hooks used to create API Gateway integrations.
+ */
+export interface ApiGatewayVtlIntegration {
+  /**
+   * Render the Request Payload as a VTL string.
+   */
+  renderRequest: (call: CallExpr, context: APIGatewayVTL) => string;
+
+  /**
+   * Construct an API GW integration.
+   */
+  createIntegration: (
+    options: aws_apigateway.IntegrationOptions & {
+      credentialsRole: aws_iam.IRole;
+    }
+  ) => aws_apigateway.Integration;
+}
+
+export interface LambdaProxyApiIntegrationProps
+  extends Omit<
+    aws_apigateway.LambdaIntegrationOptions,
+    | "requestParameters"
+    | "requestTemplates"
+    | "integrationResponses"
+    | "passthroughBehavior"
+    | "proxy"
+  > {
+  function: Function<APIGatewayProxyEvent, APIGatewayProxyResult>;
+  httpMethod: HttpMethod;
+  resource: aws_apigateway.IResource;
+}
+
+export class LambdaProxyApiMethod extends BaseApiIntegration {
+  readonly function;
+  readonly method;
+
+  constructor(private readonly props: LambdaProxyApiIntegrationProps) {
+    super();
+    this.function = props.function;
+
+    this.method = props.resource.addMethod(
+      props.httpMethod,
+      new aws_apigateway.LambdaIntegration(this.function.resource, {
+        ...this.props,
+        proxy: true,
+      })
+    );
+  }
+}
+
+export type ExcludeNullOrUndefined<T> = T extends undefined | null | infer X
+  ? X
+  : T;
+
+type IfNever<T, Default> = T extends never ? Default : T;
+
+export type Params<Request extends ApiRequest> = Exclude<
+  IfNever<
+    ExcludeNullOrUndefined<Request["path"]> &
+      ExcludeNullOrUndefined<Request["query"]>,
+    ApiParameters
+  >,
+  undefined
+>;
+
+/**
+ * The `$input` VTL variable containing all of the request data available in API Gateway's VTL engine.
+ *
+ * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html#input-variable-reference
+ */
+export interface ApiGatewayInput<Request extends ApiRequest> {
+  /**
+   * Parsed form of the {@link body}
+   */
+  readonly data: Request["body"];
+  /**
+   * The raw request payload as a string.
+   */
+  readonly body: string;
```

Comment:
Tossing and turning on this. Is there any benefit to exposing `body: string`? It is usually used as a way to render the input `body` JSON directly into the VTL, but in TS, it is used as a quoted string. I expect people to use `data` instead, perhaps we should just drop the `body` and rename `data` to `body.

Reply:
`request.data` is `$input.json("$")`